From a3815b7d8fdd2089b373b717f27fa22dd0e9dd7b Mon Sep 17 00:00:00 2001
From: Mark Stemm <mark.stemm@sysdig.com>
Date: Thu, 25 May 2017 12:24:44 -0700
Subject: [PATCH] Set close-on-exec flag for driver fds. (#851)

While debugging some agent unit tests, I noticed that after a fork +
exec the number of driver references was higher than expected. Some of
these problems can be fixed within the tests, but it's also useful to
set the close-on-exec flag for the driver fd so it's closed during an
exec.

Also fix some error messages that were referring to a hardcoded
sysdig_probe instead of PROBE_DEVICE_NAME.
---
 userspace/libscap/scap.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/userspace/libscap/scap.c b/userspace/libscap/scap.c
index 555593275..3e1c4704e 100644
--- a/userspace/libscap/scap.c
+++ b/userspace/libscap/scap.c
@@ -59,7 +59,7 @@ scap_t* scap_open_live_int(char *error,
 static uint32_t get_max_consumers()
 {
 	uint32_t max;
-	FILE *pfile = fopen("/sys/module/sysdig_probe/parameters/max_consumers", "r");
+	FILE *pfile = fopen("/sys/module/" PROBE_DEVICE_NAME "_probe/parameters/max_consumers", "r");
 	if(pfile != NULL)
 	{
 		int w = fscanf(pfile, "%"PRIu32, &max);
@@ -203,7 +203,7 @@ scap_t* scap_open_live_int(char *error,
 			else if(errno == EBUSY)
 			{
 				uint32_t curr_max_consumers = get_max_consumers();
-				snprintf(error, SCAP_LASTERR_SIZE, "Too many sysdig instances attached to device %s. Current value for /sys/module/sysdig_probe/parameters/max_consumers is '%"PRIu32"'.", filename, curr_max_consumers);
+				snprintf(error, SCAP_LASTERR_SIZE, "Too many sysdig instances attached to device %s. Current value for /sys/module/" PROBE_DEVICE_NAME "_probe/parameters/max_consumers is '%"PRIu32"'.", filename, curr_max_consumers);
 			}
 			else
 			{
@@ -214,6 +214,13 @@ scap_t* scap_open_live_int(char *error,
 			return NULL;
 		}
 
+		// Set close-on-exec for the fd
+		if (fcntl(handle->m_devs[j].m_fd, F_SETFD, FD_CLOEXEC) == -1) {
+			snprintf(error, SCAP_LASTERR_SIZE, "Can not set close-on-exec flag for fd for device %s (%s)", filename, strerror(errno));
+			scap_close(handle);
+			return NULL;
+		}
+
 		//
 		// Map the ring buffer
 		//
